### PR TITLE
crud http

### PR DESCRIPTION
**Pull request by atharva29**

### CODE REVIEW

File: api/db/batch.go [hunk 1]
```diff
@@ -45,7 +44,7 @@ func (b *Batch) Add(request *Request) (error) {
 	if b.config.Token != req.getToken() {
 		return errors.New("Cannot Batch Requests using Different Tokens")
 	}
-	allReq := &proto.AllRequest{}
+	allReq := model.AllRequest{}
 	allReq.Col = req.getCollection()
 	allReq.Operation = req.getOperation()
 	if req, ok := req.(*Insert); ok {
```
Comment (YourTechBud) [inline api/db/batch.go]: Can we use a type switch here instead?

File: api/db/batch.go [hunk 2]
```diff
@@ -57,24 +56,12 @@ func (b *Batch) Add(request *Request) (error) {
...
-		if err != nil {
-			return err
-		}
-		allReq.Find = findJSON
-		updateJSON, err := json.Marshal(req.getUpdate())
-		if err != nil {
-			return err
-		}
-		allReq.Update = updateJSON
+		allReq.Find = map[string]interface{}{"find": req.getFind()}
```
Comment (YourTechBud) [inline api/db/batch.go]:
```suggestion
		allReq.Find = req.getFind()
```

File: api/db/batch.go [hunk 3]
```diff
@@ -57,24 +56,12 @@ func (b *Batch) Add(request *Request) (error) {
...
+		allReq.Update = map[string]interface{}{"update": req.getUpdate()}
```
Comment (YourTechBud) [inline api/db/batch.go]:
```suggestion
		allReq.Update = req.getUpdate()
```

File: api/db/batch.go [hunk 4]
```diff
@@ -57,24 +56,12 @@ func (b *Batch) Add(request *Request) (error) {
...
 		allReq.Type = utils.Update
 	}
 	if req, ok := req.(*Delete); ok {
-        findJSON, err := json.Marshal(req.getFind())
-		if err != nil {
-			return err
-		}
-		allReq.Find = findJSON
+		allReq.Find = map[string]interface{}{"find": req.getFind()}
```
Comment (YourTechBud) [inline api/db/batch.go]:
```suggestion
		allReq.Find = req.getFind()
```

File: api/db/batch.go (hunk 1 above)
Comment (YourTechBud) [reply api/db/batch.go]: Do not marshal the doc and insert, won't work, put it directly

File: api/model/meta.go [hunk 5]
```diff
@@ -1,8 +1,3 @@
 package model
 
 // Meta contains the meta information required to make a request
-type Meta struct {
```
Comment (YourTechBud) [inline api/model/meta.go:4]: Delete this file

File: api/transport/aggr.go [hunk 6]
```diff
@@ -2,28 +2,24 @@ package transport
...
-	if err != nil {
-		return nil, err
-	}
+func (t *Transport) Aggr(ctx context.Context, meta *model.Meta, a *model.AggregateRequest) (*model.Response, error) {
+
+	url := t.generateDatabaseURL(meta, utils.Aggregate)
 
-	req := proto.AggregateRequest{Pipeline: pipelineJSON, Meta: meta, Operation: op}
-	res, err := t.stub.Aggregate(ctx, &req)
+	status, result, err := t.makeHTTPRequest(meta.Token, url, utils.M{"pipe": a.Pipeline, "op": a.Operation})
```
Comment (YourTechBud) [inline api/transport/aggr.go]:
```suggestion
	status, result, err := t.makeHTTPRequest(meta.Token, url, a)
```

File: api/transport/batch.go [hunk 7]
```diff
@@ -2,22 +2,24 @@ package transport
...
 
 // Batch triggers the gRPC batch function on space cloud
-func (t *Transport) Batch(ctx context.Context, meta *proto.Meta, requests []*proto.AllRequest) (*model.Response, error) {
-	req := proto.BatchRequest{Meta: meta, Batchrequest: requests}
-	res, err := t.stub.Batch(ctx, &req)
+func (t *Transport) Batch(ctx context.Context, meta *model.Meta, r *model.BatchRequest) (*model.Response, error) {
+	url := t.generateDatabaseURL(meta, utils.Batch)
+
+	// Fire the http request
+	status, result, err := t.makeHTTPRequest(meta.Token, url, utils.M{"reqs": r.Requests})
```
Comment (YourTechBud) [inline api/transport/batch.go]:
```suggestion
	status, result, err := t.makeHTTPRequest(meta.Token, url, r)
```

File: api/transport/delete.go [hunk 8]
```diff
@@ -8,13 +8,13 @@ import (
...
 // Delete triggers the gRPC delete function on space cloud
-func (t *Transport) Delete(ctx context.Context, meta *model.Meta, op string, find utils.M) (*model.Response, error) {
+func (t *Transport) Delete(ctx context.Context, meta *model.Meta, d *model.DeleteRequest) (*model.Response, error) {
 
 	// Make url for request
 	url := t.generateDatabaseURL(meta, utils.Delete)
 
 	// Fire the http request
-	status, result, err := t.makeHTTPRequest(meta.Token, url, utils.M{"find": find, "op": op})
+	status, result, err := t.makeHTTPRequest(meta.Token, url, utils.M{"find": d.Find, "op": d.Operation})
```
Comment (YourTechBud) [inline api/transport/delete.go]:
```suggestion
	status, result, err := t.makeHTTPRequest(meta.Token, url, d)
```

File: api/transport/get.go [hunk 9]
```diff
@@ -2,29 +2,24 @@ package transport
...
-	if err != nil {
-		return nil, err
-	}
+func (t *Transport) Read(ctx context.Context, meta *model.Meta, r *model.ReadRequest) (*model.Response, error) {
+	url := t.generateDatabaseURL(meta, utils.Read)
 
-	req := proto.ReadRequest{Find: findJSON, Meta: meta, Operation: op, Options: options}
-	res, err := t.stub.Read(ctx, &req)
+	// Fire the http request
+	status, result, err := t.makeHTTPRequest(meta.Token, url, utils.M{"find": r.Find, "op": r.Operation, "options": &r.Options})
```
Comment (YourTechBud) [inline api/transport/get.go]:
```suggestion
	status, result, err := t.makeHTTPRequest(meta.Token, url, r)
```

File: api/transport/insert.go [hunk 10]
```diff
@@ -2,28 +2,26 @@ package transport
...
-		return nil, err
-	}
+func (t *Transport) Insert(ctx context.Context, meta *model.Meta, i *model.CreateRequest) (*model.Response, error) {
+	// Make url for request
+	url := t.generateDatabaseURL(meta, utils.Create)
 
-	req := proto.CreateRequest{Document: objJSON, Meta: meta, Operation: op}
-	res, err := t.stub.Create(ctx, &req)
+	// Fire the http request
+	status, result, err := t.makeHTTPRequest(meta.Token, url, utils.M{"doc": i.Document, "op": i.Operation})
```
Comment (YourTechBud) [inline api/transport/insert.go]:
```suggestion
	status, result, err := t.makeHTTPRequest(meta.Token, url, i)
```

File: api/transport/update.go [hunk 11]
```diff
@@ -2,34 +2,24 @@ package transport
...
-	if err != nil {
-		return nil, err
-	}
+func (t *Transport) Update(ctx context.Context, meta *model.Meta, d *model.UpdateRequest) (*model.Response, error) {
+	url := t.generateDatabaseURL(meta, utils.Update)
 
-	req := proto.UpdateRequest{Find: findJSON, Update: updateJSON, Meta: meta, Operation: op}
-	res, err := t.stub.Update(ctx, &req)
+	// Fire the http request
+	status, result, err := t.makeHTTPRequest(meta.Token, url, utils.M{"find": d.Find, "op": d.Operation, "update": d.Update})
```
Comment (YourTechBud) [inline api/transport/update.go]:
```suggestion
	status, result, err := t.makeHTTPRequest(meta.Token, url, d)
```

**Review by YourTechBud — CHANGES_REQUESTED** [review]

File: api/db/batch.go (hunk 4 above)
Comment (atharva29) [reply api/db/batch.go]: i have changed the AllRequest struct ,in that i changed Find & Delete field from map[string]interface{} to interface only

**Comment by YourTechBud** [comment]
Closing this PR since we have decided to continue with the gRPC library

**Pull request merged**